### ISSUE TITLE
Update to Chokidar v3 and bundle it as the default

### ISSUE THREAD

**Issue by paulmillr** [rollup/rollup#2988]
I suggest to update to chokidar v3.

Chokidar v3 incorporates a lot of improvements. We've decreased RAM consumption on large directories by a huge amount; and reduced dependency count by a factor of 13. All releases are signed now. See my blog post: [Chokidar 3: How to save 32TB of traffic every week with one NPM package](https://paulmillr.com/posts/chokidar-3-save-32tb-of-traffic/)

Chokidar v3 requires nodejs 8 and higher. Is it cool with Rollup?

**Comment by lukastaegert** [comment]
Rollup only has a devDependency on chokidar and it is not bundled. You should be able to install any version you like. If v3 does not work, then that is worth another ticket. The reason v2 is a devDependency is of course because Node 6 tests would not run with v3. The long term trend, though, is probably to move to another watch solution anyway.

The next major version of Rollup will switch to a Node 8 baseline.

**Comment by paulmillr** [comment]
> The long term trend, though, is probably to move to another watch solution anyway

I don't think there's anything good out there. Node.js core wants to include parts of chokidar as a high-level simplification, but that seems to be years away.

**Comment by shellscape** [comment]
After discussion in #2707 I think it's prudent to update chokidar and make it the default. I'll be tackling that work soon.

**Comment by lukastaegert** [comment]
Still I want to make a few things clear as people seem very confused about it:
- Rollup does *not* bundle chokidar at all. It is like an optional dependency. If it is present we use it, otherwise we do not use it
- The reason we have chokidar@2 in our devDependencies is for testing only. Version 3 will work equally well if you have it installed. With the next major, we will drop Node 6 support at which point we can update our devDependency as well, but this should be of no relevance to anyone.
- Bundling chokidar would automatically transpile version 3 to something Node 6 compatible, so this will work (though I did not get it to bundle right away, there is still a lot of complexity apparently). But it will add 2 * 323kB to our 4MB bundle (once for CJS and once for ESM), which is non-trivial
- If `cheap-watch` is as reliable as `chokidar` with regard to triggering rebuilds (this is more something to be confirmed/falsified than something I have any idea about) I see no reason why we should bundle something that is 10x as big if we do not use any of its advanced features.
- Also `chokidar` is all CommomJS with lots of transitive CommonJS dependencies, so there will probably be little tree-shaking and additional wrapper overhead. Cheap-watch, on the other hand is all ESM.

I do not want to say we should not bundle chokidar, maybe cheap-watch is not capable of meeting our needs. But I want recommend to check our demands and think about the arguments above when making a call. Best to try something first, of course.

**Comment by shellscape** [comment]
I can get down with all of those points, but I do think we need to settle on one implementation and stick with it as the default. Allowing alternative watcher libs is added complexity that has shown to cause confusion. If it's not with the watcher itself, it manifests in the form of dependency resolution issues and such.

I'd like to find how much of chokidar can be tree-shaken, if at all. The userbase of `chokidar` is compelling when combined with the crazy active development and attention to detail by the org that owns it. That's a plus for Rollup WRT stability (and not to mention, security for end users).

**Comment by paulmillr** [comment]
One of the main benefits of chokidar is fsevents support on macs. 

Most developers these days use a mac. Which means they’ll benefit hugely from incorporating a native events system. This would decrease CPU and RAM usage on projects of any size.

Keep in mind that it’s likely that a project which uses rollup would also use chokidar somewhere deep in their dependency tree. Users may already have it!

**Comment by paulocoghi** [comment]
I cordially disagree with @paulmillr regarding the `fsevents` part.

I previously used **chokidar** a lot, but I'm using **cheap-watch** for almost a year in my projects without any issues, on all operating systems, including MacOS.

I can't generalize my scenario but, for me, **cheap-watch** is 100% reliable **and** my resource usage (CPU and RAM) doesn't differ after the transition to **cheap-watch**.

Of course, the difference in size between both libraries is notorious, since **cheap-watch** is **incredibly small**.

**Comment by paulmillr** [comment]
@paulocoghi don't use **bold** to emphasize your unproven claims. Do the math.

I've just tested [(see gist)](https://gist.github.com/paulmillr/a608b0fd9ad42395dc0c469de8d7de05) "cheap-watch" and compared it to chokidar.

20k files, macos-latest, node 12.8:

- chokidar: RAM: 42.07MB, CPU: 0% after initial read, took 2478ms
- cheap-watch: RAM: 1264.17MB, CPU: 70%, took 70425ms (28x slower)

That seems unusable to me.

**Comment by shellscape** [comment]
> don't use bold to emphasize your unproven claims. Do the math.

Please take that down just a notch :smile: we're all friends here. 

@paulocoghi it would be cool to get some numbers from your end to see how this compares. Some details on the systems you all are running on would be helpful too.

**Comment by paulocoghi** [comment]
I apologize for the possible misunderstanding. I only bolded the library names and size issue, when I wrote that cheap-watch is "incredibly small".

@paulmillr , I really appreciate your comprehensive test, and I also congratulate the chokidar community in reducing the library's size*. I just cordially remember that it is still 8.6x times bigger than cheap-watch.

*_As we can see from v2 to v3, the [install size reduced from 2.64MB to 330kB](https://packagephobia.now.sh/result?p=chokidar). Amazing!_

But I apologize again for not making my message clearer. I never experienced CPU or RAM problems because my projects were not that big. But **your test is really important** and it shows that cheap-watch can be improved. 

I also note that I migrated from chokidar to cheap-watch on the v2 times.

@shellscape , sure!

> it would be cool to get some numbers from your end to see how this compares

 - my main scenarios is in the usage of my front-end [framework](https://github.com/alumna/alumna/tree/2.0). I've taught about a hundred students with different operational systems, and a lot of them using Mac. Not that in version 2 of my framework, I used [chokidar](https://github.com/alumna/alumna/blob/2.0/package.json#L62).

 - besides the size issue, my biggest difficulty was to be able to package my framework through Rollup. With chokidar v2, I got several errors while trying to package chokidar with my framework, which forced me to leave it as a [dependency](https://github.com/alumna/alumna/blob/2.0/package.json#L62), which greatly increased the download and installation time and size. That must have changed in chokidar v3, but I have now switched to cheap-watch and that was the reason for my recommendation. I have changed, and I am happy.

**Comment by shellscape** [comment]
@paulmillr I'm curious about one point that @lukastaegert made:

> Also chokidar is all CommomJS with lots of transitive CommonJS dependencies, so there will probably be little tree-shaking and additional wrapper overhead. Cheap-watch, on the other hand is all ESM.

What's the chance that chokidar would move to ESM to make tree-shaking possible in this context? 

At the end of the day we want what's going to be best for the userbase, while also being responsible curators of the rollup bundle and being aware of its size.

**Comment by leeoniya** [comment]
out of curiosity, what makes commonjs less tree-shakable than ESM? i've been using rollup-plugin-cjs-es [1] with 0 issues (though i do avoid using default exports and doing dynamic/conditional `require()`s).

https://www.npmjs.com/package/rollup-plugin-cjs-es

**Comment by paulmillr** [comment]
I don’t think esm are supported by any LTS node version right now. So, it doesn’t make much sense to include it and increase bundle size.

**Comment by lukastaegert** [comment]
> out of curiosity, what makes commonjs less tree-shakable than ESM

It depends on the structure of the CJS. If exports are just made by mutating explicit properties of `exports` and there are no conditional imports or circular dependencies and ideally all `require` statements are at the top, plugins such as the one you listed or rollup-plugin-commonjs are able to transform everything safely into a tree-shakeable form.

However as soon as there are inline requires, the module needs to be wrapped into a function for conditional exection which usually causes issues for tree-shaking. Also any direct assignments to module.exports or non-resolvable mutations of `exports` will cause issues. Also there is always additional interop code to extract the correct default export. In the end it is more a hit and miss.

If you look at chokidar, there are a lot of `requires` both internal and external (there are 12 external dependencies). I did not check them all but there are at least a lot of default imports.

However the numbers provided look very convincing so we should definitely see if we can get it to bundle correctly.

**Comment by lukastaegert** [comment]
I think the ton of external dependencies (such as the ubiquitous `braces`) will prevent chokidar from providing an ESM build anyway.

**Comment by lukastaegert** [comment]
> I don’t think esm are supported by any LTS node version right now. So, it doesn’t make much sense to include it and increase bundle size

True. The only people benefitting from it would be people who bundle it, and that would be mostly us I guess. On the other hand, there is always the possibility of publishing a `chokidar-module` or `chokidar-es` package just for us 😜

**Comment by leeoniya** [comment]
maybe if it's possible for chokidar to be re-authored in a tree-shakable-after-transforms cjs format, it would work for all parties.

**Comment by paulmillr** [comment]
I'm not sure what are we talking here about. What is tree-shaking? Is it only using necessary code and ditching everything else?

I've made a huge effort to get rid of all the shit in chokidar's dependency tree — it's as minimal as it could possibly be. `braces` are there for a reason (globbing), and they were 2 megabytes before — tiny right now.

**Comment by leeoniya** [comment]
> What is tree-shaking? Is it only using necessary code and ditching everything else?

yes.

> I don't suggest creating a new chokidar fork

did someone propose this? i'm just saying that if chokidar was tweaked to avoid some patterns, like inline requires, eg. [1], [2], then it would be more tree-shakable. however, if there's a deeper chokidar dependency that contains an inline require somewhere, then those tweaks might not matter; maybe it's not even possible to avoid inline requires in chokidar.

[1] https://github.com/paulmillr/chokidar/blob/master/lib/fsevents-handler.js#L8
[2] https://github.com/paulmillr/chokidar/blob/master/lib/fsevents-handler.js#L187

**Comment by paulmillr** [comment]
[1] is unfixable, that's what `optionalDependencies` are for. Fixed [2] in https://github.com/paulmillr/chokidar/commit/c301bbe17edcbf39e3ee609a38d22b63bb4187a7.

The only place where imports aren't at the top in chokidar deps is [3], which i've [submitted PR](https://github.com/micromatch/picomatch/pull/24) for.

**Comment by lukastaegert** [comment]
Thanks @paulmillr for being so responsive about these things!

I think [1] is a real issue as this cannot be translated to an ESM build and will therefore not work in Rollup's ESM build. Incidentally, at Rollup we had a similar problem for which we used the `require-relative` package instead of `require`, which again uses the `module` builtin package to make this work.

But I think there is an even shorter, ESM compatible replacement here:
https://nodejs.org/api/modules.html#modules_module_createrequire_filename

But we could also try to inject this change from our side when bundling via a transformer plugin.

**Comment by lukastaegert** [comment]
> But I think there is an even shorter, ESM compatible replacement here

Ah, but this requires Node 12. Maybe we stick with something closer to `require-relative` then.

**Comment by paulmillr** [comment]
Again, I don't understand the end goal here. You want to replace one line of `try {require("fsevents")} catch() {}` with this construction for all 15 million weekly users of chokidar. What exact benefit would that bring? Is it going to decrease package size 2x for rollup?

**Comment by lukastaegert** [comment]
> with this construction for all 15 million weekly

probably not, as I wrote, we can do this transformation downstream. The point is, the original line prevents us from converting the code to ESM which again prevents us from bundling chokidar which is definitely our goal here, especially since chokidars dependencies have some overlap with the dependencies we already bundle.

**Comment by paulmillr** [comment]
What is the end goal of "bundling chokidar"?

**Comment by lukastaegert** [comment]
Faster startup time by avoiding the loading/parsing waterfall. More efficient variable bindings by hoisting modules into a shared scope and replacing imports by shared variables. Less code to be parsed and reduced memory footprint due to intelligent static dead-code removal. Usually very much improved install time as it is no longer necessary to pull in different dependencies. Also no need to pull in Readme files etc from transitive dependencies.

It is probably not much but that is what we do.

**Comment by shellscape** [comment]
To expand on Lukas' comment: Rollup dogfoods heavily and ships the npm module as a rollup'd bundle. If we adopt chokidar as the default watcher, we'll have to bundle it as well, in the bundle we ship.

**Comment by paulmillr** [comment]
Any updates on that? VSCode just upgraded to chokidar v3, all is good.

**Comment by lukastaegert** [comment]
It will be part of the next major. No timeframe, though.

**Comment by lukastaegert** [comment]
Important note to consider from https://github.com/rollup/rollup/issues/1666#issuecomment-536227450:

> After 2 full days on research and hacking at the source locally here's what I've concluded:
> 
> Rollup watches individual files instead of directories. This make for efficient watching but breaks with atomic writes (safe writes) in modern editors. (@BobbyBabes mentions this in #1828 (comment))
> 
> Without modifying rollup itself there are only 2 solutions right now:
> 
> 1. Disable safe writes in your editor
> 2. Add usePolling: true to the chokidar config (or equivalent setting for other watchers)
> 
> If this were to be fixed in rollup itself I believe we'd need a way to tell it to look at the directory that our files belong in. The issue here is that rather than being able to treeshake the code and know exactly what files to watch you risk watching too much (such as if you reference something in node-modules and want to watch 1 file in the module rather than the entire module itself).
> 
> I see 2 potential options:
> 
> 1. You tell rollup to blindly watch the directory of the files it wants to watch (rather than the files directly).
> 2. You manually tell rollup what directories to watch and the burden falls on you to know what you need watched.

**Comment by lukastaegert** [comment]
An approach from Rollup that I see is to actively watch directories instead of individual files and then filter the events based on which files are actually being watched. Not sure how polling works, but I would assume that that approach is more efficient.

**Comment by lukastaegert** [comment]
Maybe there is also a chokidar option to help here that does the lifting for us?

**Comment by paulmillr** [comment]
atomic: true

**Comment by paco3346** [comment]
Does the `atomic: true` option deal with the issue of inodes changing when you're watching the file itself and not the directory?

In my testing I got `atomic: true` to work if I was watching a directory but not specific files.

**Comment by paulmillr** [comment]
Not sure what are you talking about — stable chokidar versions work properly with all kinds of editors, otherwise vscode etc wouldn't use it. Some systems require usePolling: true, but that may be enabled by default on linux/win32.

**Comment by paco3346** [comment]
Sorry, I wasn't trying to suggest that there's something that's not working correctly with chokidar. In my experience the combination of rollup, chokidar, and an editor with safe writes enabled doesn't work as expected. In this case by 'expected' I mean the ability make multiple edits and still have rollup apply those changes.

My understanding of the reason this breaks with a watcher like vanilla `fs.watch()` is because it watches the inode so when a safe write editor swaps out and relinks the file `fs.watch()` is still paying attention to the original inode.

Looking at https://github.com/paulmillr/chokidar/blob/6cd588c7be458348130a2f365f563d6967010d2b/index.js#L518-L533 I see that chokidar will catch the unlink event and wait for the add event and then emit that as a `change` to the listener (assuming I'm reading that correctly).

All that being said- am I correct in thinking that if I watch `foo/bar.txt` with chokidar and then atomic swap the file with something else that my listener will get a `change` event from chokidar?

**Comment by paulmillr** [comment]
I think you'll get the event, yes

**Comment by paco3346** [comment]
It turns out that https://github.com/paulmillr/chokidar/blob/6cd588c7be458348130a2f365f563d6967010d2b/index.js#L519 never gets hit if you're on Linux and watching files directly. This is because `fs.watch()` can, by definition, never throw an `unlink` event. (I don't fully understand when `unlink` would be emitted. During a symlink change or fsevents with macOS?)

I was able to hack together a patch that keeps a copy of the original inode for each file being watched, catches `rename` events from fs.watch, compares the original inode to the one currently on disk at the given path, and if they're different, replaces the existing fs.watch instance with a new one. Essentially each `fs.watch()` instance only gets used once but it's more efficient keep recreating these than it is use polling on a large number of files.

Note that I've provided full copies of the 2 methods I modified for readability sake. The actual changes aren't too drastic.

```
function createFsWatchInstance(path, options, listener, errHandler, emitRaw) {
  const originalInode = fs.statSync(path).ino;

  const handleEvent = (rawEvent, evPath) => {
    //no good way to tell if this was deleted (and impossible to tell what it was renamed to)
    //if there's a file in place now with the name of the file we were originally watching get it's inode and compare
    //if the inodes are different that means the files were swapped (common called safe writes) atomically
    //emit a recreate event (which isn't a real fs.watch event so this should't falsely trigger anything)
    if (rawEvent === 'rename') {
      const newInode = fs.statSync(path).ino;
      if (newInode != originalInode) {
        emitRaw('recreate', evPath, {watchedPath: path});
      }
    }
    listener(path);
    emitRaw(rawEvent, evPath, {watchedPath: path});

    // emit based on events occurring for files from a directory's watcher in
    // case the file's watcher misses it (and rely on throttling to de-dupe)
    if (evPath && path !== evPath) {
      fsWatchBroadcast(
        sysPath.resolve(path, evPath), 'listeners', sysPath.join(path, evPath)
      );
    }
  };
  try {
    return fs.watch(path, options, handleEvent);
  } catch (error) {
    errHandler(error);
  }
}
```
```
const setFsWatchListener = (path, fullPath, options, handlers) => {
  const listener = handlers.listener;
  const errHandler = handlers.errHandler;
  const rawEmitter = handlers.rawEmitter;
  let cont = FsWatchInstances.get(fullPath);

  if (options.replace && cont) {
    console.log('closing existing watcher to open new one');
    cont.close();
  }

  /** @type {fs.FSWatcher=} */
  let watcher;
  if (!options.persistent) {
    watcher = createFsWatchInstance(
      path, options, listener, errHandler, rawEmitter
    );
    return watcher.close.bind(watcher);
  }
  if (cont) {
    cont.listeners.add(listener);
    cont.errHandlers.add(errHandler);
    cont.rawEmitters.add(rawEmitter);
  } else {
    let createWatcher = () => {
      let watcher = createFsWatchInstance(
        path,
        options,
        fsWatchBroadcast.bind(null, fullPath, 'listeners'),
        errHandler, // no need to use broadcast here
        fsWatchBroadcast.bind(null, fullPath, 'rawEmitters')
      );

      if (!watcher) return;
      const broadcastErr = fsWatchBroadcast.bind(null, fullPath, 'errHandlers');
      watcher.on('error', async (error) => {
        cont.watcherUnusable = true; // documented since Node 10.4.1
        // Workaround for https://github.com/joyent/node/issues/4337
        if (process.platform === 'win32' && error.code === 'EPERM') {
          try {
            const fd = await open(path, 'r');
            await close(fd);
            broadcastErr(error);
          } catch (err) {}
        } else {
          broadcastErr(error);
        }
      });
      return watcher;
    }

    watcher = createWatcher();
    if (!watcher) return;

    let recreateWatcherTimeout = null;

    //this catches the recreate event and replaces the watcher
    //this should probably be de-duped
    let recreateRawEventHandler = (event, path, options) => {
      // this is a synthesized event
      if (event === 'recreate') {
        let cont = FsWatchInstances.get(fullPath);
        cont.watcher.close();
        delete cont.watcher;
        cont.watcher = createWatcher();
      }
    };

    cont = {
      listeners: new Set([listener]),
      errHandlers: new Set([errHandler]),
      rawEmitters: new Set([rawEmitter, recreateRawEventHandler]),
      watcher: watcher
    };
    FsWatchInstances.set(fullPath, cont);
  }
  // const index = cont.listeners.indexOf(listener);

  // removes this instance's listeners and closes the underlying fs_watch
  // instance if there are no more listeners left
  return () => {
    cont.listeners.delete(listener);
    cont.errHandlers.delete(errHandler);
    cont.rawEmitters.delete(rawEmitter);
    if (!cont.listeners.size) {
      // Check to protect against issue gh-730.
      // if (cont.watcherUnusable) {
      cont.watcher.close();
      // }
      FsWatchInstances.delete(fullPath);
      FUNCTIONS.forEach(key => {
        const set = cont[key];
        set.clear();
        Object.freeze(set);
      });
      cont.watcher = null;
      Object.freeze(cont);
    }
  };
};
```

**Comment by oren** [comment]
I am new to sapper, can someone add a step by step instructions for solving the issue of 'browser not showing code change'? I am on Ubuntu. Thanks!

**Comment by oren** [comment]
Update: I use nvim and adding this line in init.vim solved it: `set backupcopy=yes`

**Comment by paulocoghi** [comment]
@paco3346 , would you mind to do a PR to Chokidar with your changes?

**Comment by paulmillr** [comment]
What are you talking about here? `unlink` event is getting called on all platforms correctly. This was the goal since v0.1.

**Comment by lukastaegert** [comment]
Solved in rollup@2.0.0

**Comment by pkit** [comment]
@paco3346 

> Note that I've provided full copies of the 2 methods I modified for readability sake.

Code has changed so these are not as readable...
Three questions:

1. `options.replace` what for? Is it just always replacing everything?
2. `recreateWatcherTimeout` is dangling? Was there meant to be a timeout?
3. `new Set()` in `cont` what's the meaning? At least current `chokidar` code handles it in either case.
